ConsumerAffairs/django-urlographer
# Update AMP URLMap status code when main URLMap changes status

Functionality Required
--
When the non-amp page version status code is changed to something other than 200 (301, 302, 410), we want the status code of the amp version to be changed too, so that visitors don't see content that we have removed and Google will stop crawling and remove the pages from the index.

Acceptance Criteria
--
1. Users can still manually (via admin) set an amp page to 301, 302, or 410.
2. AMP urls work similar to the non-amp version
  * IF the canonical (non-amp version) is set to 301, 302, 410 AND the AMP page is HTTP200 THEN the AMP page should be set to 301, 302, 410 as well.
3. An admin "History" entry {{LogEntry}} should be entered for the AMP URLMap when it is changed automatically; user for this entry will be the same user as the one updating the main canonical URLMap

Practical Tests Setup
--
Initial step:
* install the latest `urlographer` in your virtualenv by:
  1. loading your main application's virtual environment
  2. `pip uninstall django-urlographer`
  3. navigate to the this urlographer PR's branch locally
  4. `python setup.py develop` -- this puts this PR's version of urlographer in your virtualenv


Test Instructions
--
* Find a published article, PR author tested with article at `/admin/canews/article/58194/change/`
* Open links to:
  1. its `URLMap`
  2. its `AMP URLMap`

* update the `URLMap` status code to 401
* navigate to the browser tab with the `AMP URLMap`

- [ ] you still see a `200` status code

* refresh the `AMP URLMap`

- [ ] `AMP URLMap` status code has been updated to 401
* click admin `History` button for AMP URLMap
- [ ] You see a message explaining how the AMP URLMap's status code was changed, similar to this:
<img width="1323" alt="screen shot 2017-03-07 at 15 51 02" src="https://cloud.githubusercontent.com/assets/1561096/23661793/f21469ec-034d-11e7-8f27-e154592798b6.png">

Now let's test a change that includes a redirect status code
* Pick up another published article (with status code 200 again), and open tabs with it's main non-AMP URLMap and its AMP URLMap as done above
* go to main URLMap
* change status code from `410` to either `301` or `302` and set it to redirect to "any" other URLMap that actually is on sitemap (use admin list filter to filter to URLMaps that are on sitemap only)
* save the main non-AMP URLMap
* navigate to the browser tab with the `AMP URLMap`

- [ ] you still see a `410` status code

* refresh the `AMP URLMap`

- [ ] `AMP URLMap` status code has been updated to `301` or `302` (depending on what you set on the main URLMap)
- [ ] `AMP URLMap` redirects to URL as set on main URLMap
* click admin `History` button for AMP URLMap
- [ ] You see a message explaining how the AMP URLMap's status code was changed, similar to this:
<img width="1252" alt="screen shot 2017-03-07 at 16 03 55" src="https://cloud.githubusercontent.com/assets/1561096/23662398/b6e52c60-034f-11e7-833a-f416593e8035.png">


That's all, thanks!
